src/gentropy/common/harmonise.py

```diff
@@ -0,0 +1,180 @@
+"""Variant harmonisation utilities."""
+
+import pyspark.sql.functions as f
+import pyspark.sql.types as t
+from pyspark.sql import DataFrame, SparkSession
+
+from gentropy.common.spark_helpers import neglog_pvalue_to_mantissa_and_exponent
+
+
+def harmonise_summary_stats(
+    spark: SparkSession,
+    raw_summary_stats_path: str,
+    tmp_variant_annotation_path: str,
+    chromosome: str,
+    colname_position: str,
+    colname_allele0: str,
+    colname_allele1: str,
+    colname_a1freq: str,
+    colname_info: str,
+    colname_beta: str,
+    colname_se: str,
+    colname_mlog10p: str,
+    colname_n: str,
+) -> DataFrame:
+    """Ingest and harmonise the summary stats.
+
+    1. Rename chromosome 23 to X.
+    2. Filter out low INFO rows.
+    3. Filter out low frequency rows.
+    4. Assign variant types.
+    5. Create variant ID for joining the variant annotation dataset.
+    6. Join with the Variant Annotation dataset.
+    7. Drop bad quality variants.
+
+    Args:
+        spark (SparkSession): Spark session object.
+        raw_summary_stats_path (str): Input raw summary stats path.
+        tmp_variant_annotation_path (str): Input variant annotation dataset path.
+        chromosome (str): Which chromosome to process.
+        colname_position (str): Column name for position.
+        colname_allele0 (str): Column name for allele0.
+        colname_allele1 (str): Column name for allele1.
+        colname_a1freq (str): Column name for allele1 frequency (optional).
+        colname_info (str): Column name for INFO, reflecting variant quality (optional).
+        colname_beta (str): Column name for beta.
+        colname_se (str): Column name for beta standard error.
+        colname_mlog10p (str): Column name for -log10(p).
+        colname_n (str): Column name for the number of samples.
+
+    Returns:
+        DataFrame: A harmonised summary stats dataframe.
+    """
+    # Read the precomputed variant annotation dataset.
+    va_df = (
+        spark
+        .read
+        .parquet(tmp_variant_annotation_path)
+        .filter(f.col("vaChromosome") == ("X" if chromosome == "23" else chromosome))
+        .persist()
+    )
+
+    # Read and process the summary stats dataset.
+    df = (
+        spark
+        .read
+        .parquet(raw_summary_stats_path)
+        .filter(f.col("chromosome") == chromosome)
+        # Harmonise, 1: Rename chromosome 23 to X.
+        .withColumn(
+            "chromosome",
+            f.when(
+                f.col("chromosome") == "23", "X"
+            ).otherwise(f.col("chromosome"))
+        )
+    )
```

Comment:
I can see it's somewhat common in ukb, mayby extact it as a helper function and add a short example?

Reply:
The current approach I've taken is for the `harmonise_summary_stats` function is to keep it monolithic, but configurable on column names when necessary. I could extract this as a helper function; however, for consistency, I think it would make sense to break down this entire function into separate smaller functions. I think that right now, given time constraints on handing everything over, it's better to not do it. So if it's all right with you, shall I keep it as is?

Reply:
Yes, lets keep it as is for now. It is not that crusial.